### PR TITLE
A plan to remove DWR

### PR DESCRIPTION
**Pull request by wweellddeerr**
Remove the DWR library from SUMA.

See the [rendered version](https://github.com/uyuni-project/uyuni-rfc/blob/6b8deb6118d29b67b49f76dfdf7c664a329d7d88/accepted/00092-replace-dwr.md).

### CODE REVIEW

File: accepted/0000-replace-dwr.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  * <sup>*</sup> `SystemGroupsRenderer`: exposes the `renderAsync` method of the [`SystemGroupsRenderer`](https://github.com/uyuni-project/uyuni/blob/master/java/code/src/com/redhat/rhn/frontend/action/renderers/SystemGroupsRenderer.java) class. It is [loaded](https://github.com/uyuni-project/uyuni/blob/master/java/code/webapp/WEB-INF/pages/yourrhn.jsp#L30) and [used](https://github.com/uyuni-project/uyuni/blob/master/java/code/webapp/WEB-INF/pages/yourrhn.jsp#L85) only in the `yourrhn.jsp` file.
+  * <sup>*</sup> `TasksRenderer`: exposes the `renderAsync` method of the [`TasksRenderer`](https://github.com/uyuni-project/uyuni/blob/master/java/code/src/com/redhat/rhn/frontend/action/renderers/TasksRenderer.java) class. It is [loaded](https://github.com/uyuni-project/uyuni/blob/master/java/code/webapp/WEB-INF/pages/yourrhn.jsp#L15) and [used](https://github.com/uyuni-project/uyuni/blob/master/java/code/webapp/WEB-INF/pages/yourrhn.jsp#L48) only in the `yourrhn.jsp` file.
+
+  <sup>*</sup> All of these are only used in the [Your RHN](https://github.com/uyuni-project/uyuni/blob/master/java/code/webapp/WEB-INF/pages/yourrhn.jsp) page.
```
Comment (moio) [inline accepted/0000-replace-dwr.md:42]: Historical note FYI: Your RHN was envisioned to be a customizable homepage with "widgets". What actually came to be is a limited set of such "widgets" and, to my knowledge, basically nobody ever changes them.

I would not object a redesign of the whole homepage, although that goes above and beyond the intent of this PR.

File: accepted/0000-replace-dwr.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+For the frontend side, the capabilities provided by DWR (both the utility functions from `util.js` file and the ajax calls) are replaceable using jQuery.
+
+There is a PoC in [this commit](https://github.com/wweellddeerr/uyuni/commit/9ddb68030dc49fa3272a062ceb7015c35b68d457) exploring how it is possible to use jQuery to make the ajax request and a raw Servlet as backend replacing the `DWRItemSelector` component of DWR. The SSM (one of the features that uses it) seems to work well after applying this patch.
+
+Using a raw Servlet as backend seems to be the best approach since (a) the methods currently exposed by DWR are only available through the `DwrServlet`, which is a raw servlet; and (b) these methods typically manipulate the HttpSession, making it difficult to use the HTTP API, for example.
```
Comment (moio) [inline accepted/0000-replace-dwr.md:56]: Can you elaborate on this comment a bit, pointing to what exactly we would need to manipulate and why?

File: accepted/0000-replace-dwr.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+# Alternatives
+
+The first alternative is to just keep using DWR and try to solve the warning mentioned above, maybe trying to create a Pull Request on the DWR repository or even opening an issue on it.
+
+The second alternative is to gradually replace the pages using DWR to use React. The main drawbacks of this alternative are: (1) some issues to migrate pages using DWR were aborted in the past, probably due to non technical decisions; and (2) the `rl:selectablecolumn` custom tag mentioned before is used in several JSP pages, it would be necessary to find an alternative to this component in order to avoid the need to migrate so many pages.
+
+A hybrid approach could also be used, migrating some pages (specially the `Your RHN`) to React and applying the suggested solution (or even the first alternative) only in the remaining cases, reducing the risks.
```
Comment (moio) [inline accepted/0000-replace-dwr.md:70]: Ultimately we will want to port all pages to React, but I think the subgoal of getting rid of DWR is an important one as well.

Please use your best judgement to select the cheapest approach - if migrating a page to React is easier than adapting the corresponting DWR methods, please go down that path.

**Review by moio — COMMENTED** [review]
Apart from a question inline I like this proposal and I think it goes in the right direction.

This is a long-time tech debt item I'd love to see tackled.

**Review by rjmateus — APPROVED** [review]
This looks like a great approach to me, to be used as a transitional step before we get rid of JSP pages (someday in the future).
In terms of risk, the most critical part is the selector one. But from a server point of view, the call will be the same, only the communication style will change. That said, it should work or fail catastroficly.

**Comment by admd** [comment]
@cbbayburt @Etheryte @cbosdo can you please gives a thumbs up to Welder if you have nothing against this approach so @wweellddeerr can proceed further?

**Review by Etheryte — APPROVED** [review]
I'm not all that familiar with the current ins and outs of DWR, but the proposed approach to replace it looks good to me. 👍

**Comment by cbosdo** [comment]
@wweellddeerr There have been a successful attempt at replacing a page using DWR feature: the virtual systems list. The new systems list also gets rid of quite a number of places using the `DWRItemSelector`. We now have an API endpoint to add items in the SSM: https://github.com/uyuni-project/uyuni/blob/master/java/code/src/com/suse/manager/webui/controllers/SetController.java#L76.

I like your approach, but I fear the temporary socket would stay for ever... We would still need to remember to progressively convert to proper API endpoints.

File: accepted/0000-replace-dwr.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+* **Dynamically generated JS files** - the following js files are generated dynamically by DWR, exposing some methods of Java classes as pure js functions that call the real code using Ajax requests. The classes and methods are defined in [`dwr.xml`](https://github.com/uyuni-project/uyuni/blob/master/java/code/webapp/WEB-INF/dwr.xml) file: 
+  * `DWRItemSelector`: exposes the method [`select`](https://github.com/uyuni-project/uyuni/blob/master/java/code/src/com/redhat/rhn/frontend/taglibs/DWRItemSelector.java#L52) of the DWRItemSelector java class. It is loaded in [`layout_head.jsp`](https://github.com/uyuni-project/uyuni/blob/master/java/code/webapp/WEB-INF/decorators/layout_head.jsp#L69) and referenced in [template.js](https://github.com/uyuni-project/uyuni/blob/master/web/html/src/styleguide/template.js#L33), so loaded in all pages. The function is currently used in all the pages using the `rl:selectablecolumn` custom tag defined [here](https://github.com/uyuni-project/uyuni/blob/master/java/code/src/com/redhat/rhn/frontend/taglibs/rhn-list.tld#L242), including SSM. **This seems to be the most risky component, since it is used across several pages**.
```
Comment (cbosdo) [inline accepted/0000-replace-dwr.md:29]: This one is used for the SSM Count. For sure it is used in a lot of pages, but as long as we find a proper way to implement it using API endpoints it should be easy to propagate to React pages at least.

File: accepted/0000-replace-dwr.md (hunk 1 above)
Comment (cbosdo) [reply accepted/0000-replace-dwr.md:42]: reactifying YourRHN would be really cool and kick a number of DWR uses: two birds killed with one stone.

**Pull request merged**